node/network/bridge/src/validator_discovery.rs

```diff
@@ -115,39 +114,93 @@ impl NonRevokedConnectionRequestState {
-	// the `u64` counts the number of pending non-revoked requests for this validator
+	// Peers that are connected to us and authority ids associated to them.
+	connected_peers: HashMap<PeerId, HashSet<AuthorityDiscoveryId>>,
+	// The `u64` counts the number of pending non-revoked requests for this validator
 	// note: the validators in this map are not necessarily present
 	// in the `connected_validators` map.
 	// Invariant: the value > 0 for non-revoked requests.
 	requested_validators: HashMap<AuthorityDiscoveryId, u64>,
 	non_revoked_discovery_requests: Vec<NonRevokedConnectionRequestState>,
 	// PhantomData used to make the struct generic instead of having generic methods
-	network: PhantomData<N>,
-	authority_discovery: PhantomData<AD>,
+	_phantom: PhantomData<(N, AD)>,
 }
 
 impl<N: Network, AD: AuthorityDiscovery> Service<N, AD> {
 	pub fn new() -> Self {
 		Self {
-			connected_validators: HashMap::new(),
+			connected_peers: HashMap::new(),
 			requested_validators: HashMap::new(),
 			non_revoked_discovery_requests: Vec::new(),
-			network: PhantomData,
-			authority_discovery: PhantomData,
+			_phantom: PhantomData,
+		}
+	}
+
+	/// Find connected validators using the given `validator_ids`.
+	///
+	/// Returns a [`HashMap`] that contains the found [`AuthorityDiscoveryId`]'s and their associated [`PeerId`]'s.
+	async fn find_connected_validators(
+		&mut self,
+		validator_ids: &[AuthorityDiscoveryId],
+		authority_discovery_service: &mut AD,
+	) -> HashMap<AuthorityDiscoveryId, PeerId> {
+		let mut result = HashMap::new();
+
+		for id in validator_ids {
+			// First check if we already cached the validator
+			if let Some(pid) = self.connected_peers
+				.iter()
+				.find_map(|(pid, ids)| if ids.contains(&id) { Some(pid) } else { None }) {
+				result.insert(id.clone(), pid.clone());
+				continue;
+			}
+
+			// If not ask the authority discovery
+			if let Some(addresses) = authority_discovery_service.get_addresses_by_authority_id(id.clone()).await {
+				for peer_id in addresses.iter().filter_map(peer_id_from_multiaddr) {
+					if let Some(ids) = self.connected_peers.get_mut(&peer_id) {
+						ids.insert(id.clone());
+						result.insert(id.clone(), peer_id.clone());
+						continue;
+					}
+				}
+			}
 		}
+
+		result
 	}
 
 	/// On a new connection request, a priority group update will be issued.
 	/// It will ask the network to connect to the validators and not disconnect
 	/// from them at least until all the pending requests containing them are revoked.
 	///
 	/// This method will also clean up all previously revoked requests.
-	// it takes `network_service` and `authority_discovery_service` by value
-	// and returns them as a workaround for the Future: Send requirement imposed by async fn impl.
+	/// it takes `network_service` and `authority_discovery_service` by value
+	/// and returns them as a workaround for the Future: Send requirement imposed by async fn impl.
```

Comment:
this is an implementation detail which I don't think should be part of docs, maybe there is a way to avoid that, but that was a quick fix I went with

Reply:
This will not be part of the docs anyway, because this is a private struct.